### PR TITLE
add a config feature in kube - fixes #533

### LINKED ISSUE #533: create a `config` feature in kube
We currently have a "private" `__config` feature in `kube` to track the config related dependencies.

I would like to expose this properly so that people can use `kube` with only the `config` feature to deal with kubeconfig related parsing.

This would require:
- having checks for `#[cfg(feature = "config")` for the `config` module
- having `config` feature depend on internal `__non_core` feature
- having checks for `#[cfg(feature = "config")` or `#[cfg(feature = "client")` (or possibly `feature = __non_core`?) for overlapping deps
- having the error enum be more feature flagged (currently entire module is gated on `client` feature)
- a small test example that parser kubeconfig without features (needs some feature-re-export hackery in examples/Cargo.toml, plus a circleci test line)

You should not be able to get the `client` feature without the `config` feature (already ensured by having `__config` feature listed as a dependency of the `client` feature), but the other way around should work.

### PR DESCRIPTION
**Pull request by clux**
surprisingly simple to split up these features

### CODE REVIEW

File: kube/src/lib.rs [hunk 1]
```diff
@@ -101,23 +101,46 @@ macro_rules! cfg_client {
...
+    ($($item:item)*) => {
+        $(
+            #[cfg_attr(docsrs, doc(cfg(feature = "config")))]
+            #[cfg(feature = "config")]
+            $item
+        )*
+    }
+}
+
+macro_rules! cfg_error {
```
Comment (kazk) [inline kube/src/lib.rs:114]: I think we should limit `cfg_` macros to use the names of feature sets. This reads like if `error` feature is enabled.

`cfg_client_or_config`?

Comment (clux) [reply kube/src/lib.rs:114]: I was going by the pattern used by tokio to have the macro name be the more useful one, and it internally selected on the correct feature set: https://github.com/tokio-rs/tokio/blob/ce9ca45c92b9bd4b449f94339fbe60a4ded9ab28/tokio/src/macros/cfg.rs

Comment (kazk) [reply kube/src/lib.rs:114]: Oh, I see. Then this is fine.
I still think we should split `kube::Error` into smaller ones, but that's a separate issue.

Comment (clux) [reply kube/src/lib.rs:114]: Yeah, more focused errors along our previous discussion is still worth doing.

**Comment by clux** [comment]
released in 0.56

**Pull request merged**